### ISSUE TITLE
[Modern] Difficulty following paginationContainer example

### ISSUE THREAD

**Issue by AndrewIngram** [facebook/relay#1625]
I'm trying to follow the paginationContainer example, but have a couple of issues:

Firstly, it won't compile:

```
Invariant Violation: RelayConnectionTransform: Expected type `EventConnection` to have an edges.cursor field for which the type is an scalar in document `UpcomingEvents_query`.
```

My EventConnection looks like this

```graphql
type EventConnection {
  pageInfo: PageInfo!
  edges: [EventEdge]
}

type EventEdge {
  node: Event
  cursor: String!
}
```

This is unchanged from my old classic-compatible schema.

Secondly, the example specifies the type of the cursor variable as "ID", but the connection spec (as well as graphql-relay-js) defines cursors as "String"

**Comment by AndrewIngram** [comment]
There's also no examples of the correct way to expose variables defined on the query in the QueryRenderer to the fragment and pagination containers, I'm assuming they'll just available, but I haven't actually managed to get it to compile yet :)

**Comment by AndrewIngram** [comment]
Okay, so it looks like the issue is in `RelayConnectionTransform` in the `validateConnectionType` function. When validating cursor and pageInfo, as well as the fields on pageInfo, there are missing calls to `RelaySchemaUtils.getNullableType`, so it's failing when these fields are marked as required (which is what `graphql-relay-js` spits out). When I manually edit my schema to loosen the restriction, compilation succeeds.

**Comment by kassens** [comment]
This is something we should allow, feel free to send a PR to add the getNullableType if you get to it before we do!

**Comment by janicduplessis** [comment]
Fixed the compilation issue in #1637

Managed to get everything working based on the example, just a few things I found out that are not clear:

- You must fetch `cursor` and some fields in `pageInfo`.
-  I had to specify `direction` and `getConnectionFromProps`
- loadMore crashes if you don't pass a callback (#1636)

**Comment by zunder** [comment]
@janicduplessis should the field definition `pageInfo` in the query be necessary? I thought `@connection` is there in order to include this field into the fragment/query during compilation time. But trying it out myself, it looks like you still have to define
```
pageInfo {
  hasNextPage
  endCursor
}
```
in the fragment manually. Is that a known bug?

**Comment by 0xCMP** [comment]
Yea I had to include what @zunder did to get my pagination component working again. Very weird since I swore that once I had added the pageInfo and it broke (but maybe I just hadn't recompiled).

None of the examples seem to show that you need to include the pageInfo.

**Comment by akomm** [comment]
I agree on the case, that the documentation is very hard to follow, impossible if not checking the actual implementation.

I started few month ago with graphql. Have one quite complex app implemented with react and react-apollo. I like the approach and idea of relay (I think I understood it quite well). My current project should run on relay modern. I already found many things working much better **for me** with relay modern, although most of it via checking the the implementation. I understand that it takes time to write full and good docs and it is not an easy task.

When you work through the docs, you have a huge gap from QueryRenderer and fragmentContainer, which is, when you follow the logic confuses. 

- Suddenly you have directives and naming convention, which might have sort of logic, but are everything else than straight-forward (Feed_feed).

- You suddenly have fragment and query all defined for one component. Why do we need fragment in this case?

- We do not use QueryRenderer. Where does the environment come from? I checked the code: it is expected to be passed via context. But checking via devtools I see no context passed with relay, even not in child components of QueryRenderer.

- What if you paginate top level connection (products for example). You would have to have a fragment on Query type, or have to create a fragment on the ProductsConnection type, containing only the edge and pageInfo. This does not really work out with the example provided. I would find a way to "adjust" it, if the docs on the paginationContainer would only be more straight-forward in terms of "the concept", not just example code. So currently I try to learn via implementation, as the docs do not help (this worked for me so far) ;( .
